nav2_recoveries/plugins/assisted_teleop.cpp

```diff
@@ -0,0 +1,300 @@
+{
+  node_ = parent;
+  auto node = node_.lock();
+
+  logger_ = node->get_logger();
+
+  RCLCPP_INFO(logger_, "Configuring %s", name.c_str());
+
+  recovery_name_ = name;
+  tf_ = tf;
+
+  nav2_util::declare_parameter_if_not_declared(
+    node,
+    "projection_time", rclcpp::ParameterValue(1.0));
+
+  nav2_util::declare_parameter_if_not_declared(
+    node,
+    "linear_velocity_threshold_", rclcpp::ParameterValue(0.06));
+
+  nav2_util::declare_parameter_if_not_declared(
+    node,
+    "cmd_vel_topic", rclcpp::ParameterValue(std::string("cmd_vel")));
+
+  nav2_util::declare_parameter_if_not_declared(
+    node,
+    "input_vel_topic", rclcpp::ParameterValue(std::string("cmd_vel_input")));
+
+  node->get_parameter("global_frame", global_frame_);
+  node->get_parameter("robot_base_frame", robot_base_frame_);
+  node->get_parameter("transform_tolerance", transform_tolerance_);
+  node->get_parameter("projection_time", projection_time_);
+  node->get_parameter("linear_velocity_threshold_", linear_velocity_threshold_);
+  node->get_parameter("cmd_vel_topic", cmd_vel_topic_);
+  node->get_parameter("input_vel_topic", input_vel_topic_);
+
+  vel_sub_ = node->create_subscription<geometry_msgs::msg::Twist>(
+    input_vel_topic_, rclcpp::SystemDefaultsQoS(),
+    std::bind(
+      &AssistedTeleop::vel_callback,
+      this, std::placeholders::_1));
+
+  vel_pub_ = node->create_publisher<geometry_msgs::msg::Twist>(cmd_vel_topic_, 1);
+
+  costmap_sub_ = std::make_unique<nav2_costmap_2d::CostmapSubscriber>(
+    node_, "local_costmap/costmap_raw");
+
+  action_server_ = std::make_shared<ActionServer>(
+    node, recovery_name_,
+    std::bind(&AssistedTeleop::execute, this));
+
+  collision_checker_ = collision_checker;
+}
+
+void
+AssistedTeleop::vel_callback(const geometry_msgs::msg::Twist::SharedPtr msg)
+{
+  speed_x = msg->linear.x;
+  speed_y = msg->linear.y;
+  angular_vel_ = msg->angular.z;
+  if (go) {
+    if (!checkCollision()) {
+      RCLCPP_INFO(logger_, "Reducing velocity by a factor of %.2f", scaling_factor);
+    }
+    moveRobot();
+  }
+}
+
+bool
+AssistedTeleop::updatePose()
+{
+  if (!nav2_util::getCurrentPose(
+      current_pose, *tf_, global_frame_, robot_base_frame_,
+      transform_tolerance_))
+  {
+    RCLCPP_ERROR(logger_, "Current robot pose is not available.");
+    return false;
+  }
+  projected_pose.x = current_pose.pose.position.x;
+  projected_pose.y = current_pose.pose.position.y;
+  projected_pose.theta = tf2::getYaw(
+    current_pose.pose.orientation);
+  return true;
+}
+
+void
+AssistedTeleop::projectPose(
+  double speed_x, double speed_y,
+  double angular_vel_, double projection_time)
+{
+  // Project Pose by time increment
+  projected_pose.x += projection_time * (
+    speed_x * cos(projected_pose.theta));
+  projected_pose.y += projection_time * (
+    speed_y * sin(projected_pose.theta));
+  projected_pose.theta += projection_time *
+    angular_vel_;
+}
+
+bool
+AssistedTeleop::checkCollision()
```

Comment:
Rename this function, this is actually doing the main body of assisted teleop, its not just checking for collisions but also scaling velocities to avoid collisions as well.

This should return a velocity in the end to use for sending to the robot, not a boolean.